File: backend/src/Solitaire.Engine/Deck.cs
using System.Collections.Immutable;

namespace Solitaire.Engine;

/// <summary>
/// Deterministic deck construction: ordered deck, seeded shuffle, and the
/// Klondike deal. All three steps are fully specified so a TypeScript port can
/// reproduce identical deals from the same seed.
/// </summary>
public static class Deck
{
    /// <summary>Total cards in a standard deck.</summary>
    public const int Size = 52;

    /// <summary>Cards dealt to the tableau at the start of a game (1+2+...+7).</summary>
    public const int TableauDealCount = 28;

    /// <summary>
    /// The canonical unshuffled deck: card ordinal 0..51 in order
    /// (Clubs A..K, Diamonds A..K, Hearts A..K, Spades A..K). See
    /// <see cref="Card.OrdinalIndex"/>.
    /// </summary>
    public static ImmutableArray<Card> BuildOrdered()
    {
        var builder = ImmutableArray.CreateBuilder<Card>(Size);
        for (int i = 0; i < Size; i++)
        {
            builder.Add(Card.FromOrdinalIndex(i));
        }

        return builder.MoveToImmutable();
    }

    /// <summary>
    /// Returns the ordered deck shuffled with a seeded Fisher–Yates
    /// (Durstenfeld) pass.
    /// </summary>
    /// <remarks>
    /// Exact algorithm (reproduce identically in TypeScript):
    /// <code>
    /// deck = BuildOrdered()               // index 0 = Ace of Clubs, ...
    /// rng  = DeterministicRandom(seed)
    /// for i from 51 down to 1:            // inclusive
    ///     j = rng.NextInt(i + 1)          // 0 &lt;= j &lt;= i
    ///     swap deck[i], deck[j]
    /// </code>
    /// The loop runs from the last index down to 1 and swaps each element with a
    /// uniformly chosen earlier-or-equal index. <c>deck[0]</c> becomes the top of
    /// the deal (first card consumed).
    /// </remarks>
    public static ImmutableArray<Card> Shuffle(int seed)
    {
        var cards = BuildOrdered().ToArray();
        var rng = new DeterministicRandom(seed);

        for (int i = cards.Length - 1; i >= 1; i--)
        {
            int j = rng.NextInt(i + 1);
            (cards[i], cards[j]) = (cards[j], cards[i]);
        }

        return [.. cards];
    }

    /// <summary>
    /// Deals a shuffled deck into the initial Klondike layout.
    /// </summary>
    /// <remarks>
    /// Deal order (column-by-column, consuming <paramref name="shuffled"/> from
    /// the front): for column <c>c</c> in 0..6, take <c>c + 1</c> cards; the first
    /// taken sits at the bottom face-down and the last taken is face-up on top, so
    /// column <c>c</c> has <c>c</c> face-down cards and one face-up card. The
    /// remaining 24 cards become the stock with index 0 as the top (next drawn).
    /// </remarks>
    public static (ImmutableArray<TableauPile> Tableau, ImmutableArray<Card> Stock) Deal(
        ImmutableArray<Card> shuffled)
    {
        if (shuffled.Length != Size)
        {
            throw new ArgumentException("A full 52-card deck is required.", nameof(shuffled));
        }

        var tableau = ImmutableArray.CreateBuilder<TableauPile>(GameState.TableauCount);
        int next = 0;
        for (int c = 0; c < GameState.TableauCount; c++)
        {
            int cardsInColumn = c + 1;
            var column = ImmutableArray.CreateBuilder<Card>(cardsInColumn);
            for (int r = 0; r < cardsInColumn; r++)
            {
                column.Add(shuffled[next++]);
            }

            // All but the last dealt card are face-down.
            tableau.Add(new TableauPile(column.MoveToImmutable(), cardsInColumn - 1));
        }

        var stock = ImmutableArray.CreateBuilder<Card>(Size - next);
        for (int i = next; i < Size; i++)
        {
            stock.Add(shuffled[i]);
        }

        return (tableau.MoveToImmutable(), stock.MoveToImmutable());
    }
}

File: frontend/src/engine/deck.ts
// Deck construction, the shared seeded shuffle, and the Klondike deal —
// port of Deck.cs. Reproduces the documented Fisher–Yates pass exactly.

import { cardFromOrdinal, type Card } from './cards';
import { DeterministicRandom } from './random';
import type { TableauPile } from './tableau';

export const DECK_SIZE = 52;
export const KLONDIKE_TABLEAU_COUNT = 7;

/** The canonical unshuffled 52-card deck (ordinal 0..51). */
export function buildOrdered(): Card[] {
  const cards: Card[] = [];
  for (let i = 0; i < DECK_SIZE; i++) {
    cards.push(cardFromOrdinal(i));
  }
  return cards;
}

/**
 * The shared seeded Fisher–Yates (Durstenfeld) shuffle used by every variant.
 * For i from length-1 down to 1, swap items[i] with items[rng.nextInt(i + 1)].
 */
export function shuffleInPlace(items: Card[], rng: DeterministicRandom): void {
  for (let i = items.length - 1; i >= 1; i--) {
    const j = rng.nextInt(i + 1);
    const tmp = items[i]!;
    items[i] = items[j]!;
    items[j] = tmp;
  }
}

/** Returns the ordered 52-card deck shuffled with the seeded Fisher–Yates pass. */
export function shuffle(seed: number): Card[] {
  const cards = buildOrdered();
  shuffleInPlace(cards, new DeterministicRandom(seed));
  return cards;
}

/**
 * Deals a shuffled deck into the initial Klondike layout: column c (0..6) gets
 * c+1 cards, the last dealt face-up; the remaining 24 cards form the stock
 * (index 0 = top).
 */
export function dealKlondike(shuffled: readonly Card[]): {
  tableau: TableauPile[];
  stock: Card[];
} {
  const tableau: TableauPile[] = [];
  let next = 0;
  for (let c = 0; c < KLONDIKE_TABLEAU_COUNT; c++) {
    const cardsInColumn = c + 1;
    const column: Card[] = [];
    for (let r = 0; r < cardsInColumn; r++) {
      column.push(shuffled[next++]!);
    }
    tableau.push({ cards: column, faceDownCount: cardsInColumn - 1 });
  }
  return { tableau, stock: shuffled.slice(next) };
}
